8224228: No way to locally suppress lint warnings in parser/tokenizer or preview features

archiecobbs opened the pull request:
This PR updates the `DeferredLintHandler` so that deferred warnings can be registered during parsing, before any `JCTree` nodes have been created, and it uses this new capability to update the `"preview"` and `"text-blocks"` Lint warnings so that they can be suppressed via `@SuppressWarnings` annotations. More details are provided in additional comments.

Note: This PR depends on (i.e., includes and extends) these other "cleanup" PR's:
* #23167
* #23281
* #23400
* #23730

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8224228](https://bugs.openjdk.org/browse/JDK-8224228): No way to locally suppress lint warnings in parser/tokenizer or preview features (**Task** - P4)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/23237/head:pull/23237` \
`$ git checkout pull/23237`

Update a local copy of the PR: \
`$ git checkout pull/23237` \
`$ git pull https://git.openjdk.org/jdk.git pull/23237/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 23237`

View PR using the GUI difftool: \
`$ git pr show -t 23237`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/23237.diff">https://git.openjdk.org/jdk/pull/23237.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/23237#issuecomment-2608208616)
</details>

archiecobbs commented:
**Overview**

The purpose of `DeferredLintHandler` is to allow `@SuppressWarnings` to be applied to warnings that are generated before `@SuppressWarnings` annotations themselves have been processed. The way this currently works is that warning callbacks are kept in a `HashMap` keyed by the innermost containing module, package, class, method, or variable declarations (in the form of a `JCModuleDecl`, `JCPackageDecl`, `JCClassDecl`, `JCMethodDecl`, or `JCVariableDecl`). Later, when the compiler executes the attribution phase and the lint categories suppressed at each declaration are known, the corresponding warning callbacks are provided with an appropriately configured `Lint` instance and "flushed".

During lexical scanning, `JCTree` nodes don't exist yet, so it is impossible to follow this protocol. Similarly, during parsing, `JCTree` nodes are in the process of being created, so  at best, some awkward gymnastics are required, e.g., see the handling for `"dangling-doc-comments"`.

This PR refactors `DeferredLintHandler` so that it can handle warnings generated during parsing:
* During parsing, warnings are temporarily stored by lexical position
* After parsing, the lexical warnings are remapped to the innermost containing `JCTree` declaration

One side effect of this approach is that parsing must now always be done with an `EndPositionTable`, because it's required to calculate whether a lexical position is within the lexical extent of a declaration. However, after parsing this table can be discarded after parsing (if not needed). Not sure if/how this would meaningfully impact compile time.

**Per-file comments**

`Preview.java`
* Lexical warnings are now deferred via `DeferredLintHandler`
* Normal warnings via `warnPreview()` are now passed the current `Lint` instance
* Normal warnings via `checkSourceLevel()` are now passed the current `Lint` instance

`Annotate.java`
* Updated for new `DeferredLintHandler` methods for `push()` and `pop()`
* Tighten `DiagnosticPosition deferPos` to `JCTree deferDecl` which is what's actually required
* Line 680: Fix a bug where a `JExpression` (which is not a declaration supporting `@SuppressWarnings`) was being used for `deferPos`. In this case the correct `deferDecl` should already be established, so we can replace with null.

`Attr.java`
* Updated for new `DeferredLintHandler` methods for `push()` and `pop()`
* Add a couple of missed `flush()`s for variable declarations:
  * Pattern matching binding
  * Enhanced `for()` loop variables

`Check.java`
* Updated for new `DeferredLintHandler` methods for `push()` and `pop()`
* Pass current `Lint` to `preview.reportPreviewWarning()`
* No need to defer in `checkPreview()`
* Update `SuperThisChecker` to handle `@SuppressWarnings("preview")` on constructors

`Modules.java`
* Updated for new `DeferredLintHandler` methods for `push()` and `pop()`
* Update to support `@SuppressWarnings("preview")` on module declarations

`JavaCompiler.java`
* Update `DeferredLintHandler` when entering and exiting parsing
* `EndPositionTable` is now required while parsing, but can be discarded afterwards

`JavaTokenizer.java`
* `lexWarning()` can now defer warnings via `DeferredLintHandler` (this fixes `"text-blocks"`)

`JavacParser.java`
* Make `EmptyEndPosTable` public so `JavaCompiler.java` can create one

`MandatoryWarningHandler.java`
* Allow overriding `verbose` on a per-report basis to reflect `@SuppressWarnings`

Changes to existing tests:

`ImplicitClass/ErrorRecovery.java`
* The compiler error now results in elimination of the "recompile" note, which is inappropriate in light of the error

`preview/PreviewAutoSuppress.java`
* Warnings reordered as side effect of deferral, but note they are now consistent with the previous test

`preview/PreviewErrors.java`
* Updated to reflect the new effectiveness of `@SuppressWarnings("preview")`

**Other notes**

I tried to cleanup the existing logic in `JavacParser.java` for handling `"dangling-doc-comments"`(+13, -46), but it led to this test failure:
```java
    @SuppressWarnings("dangling-doc-comments")
    /// Bad/misplaced/suppressed comment.
    public void m2() { }
```
The current code applies the suppression to the misplaced comment, thereby suppressing the warning. But lexicaly speaking, that comment is not actually within the declaration of `m2()` to which the `@SuppressWarnings` annotation applies, so one could argue that the `@SuppressWarnings` should not apply to it and therefore this test is wrong. Refactoring this code results in the latter interpretation and so the test would fail.

mcimadamore commented:
Question: the change looks beneficial, but there's a lot going on in the PR. Would it be possible, do you think, to break it down into simpler parts? For instance, the push/pop logic seems something that we could review and integrate separately (and it's adding quite a bit of noise to the PR). Same for missing flushes etc.

archiecobbs commented:
Hi @mcimadamore,

Thanks for taking a look.

> Would it be possible, do you think, to break it down into simpler parts? 

Absolutely, I will work on that.

mcimadamore commented:
> Hi @mcimadamore,
> 
> Thanks for taking a look.
> 
> > Would it be possible, do you think, to break it down into simpler parts?
> 
> Absolutely, I will work on that.

Thanks!!

archiecobbs commented:
Note: Rebasing this PR on #23281 which should be reviewed first.

archiecobbs commented:
@mcimadamore,

> Would it be possible, do you think, to break it down into simpler parts?

I've split this into three precursor "cleanup" PR's (although the third one is new and wasn't part of the original change).

So this PR depends on (i.e., includes and extends) these others. Obviously they would need to be adjudicated first:
* #23167
* #23281
* #23400

Let me know if this looks like a reasonable approach. I'm happy to split things up differently if needed. Thanks.

archiecobbs commented:
> Question -- in a [related change](https://github.com/openjdk/jdk/pull/23281) we cleaned up the deferred lint handler code to use trees instead of positions, as positions were too weak. This PR now adds a parsing mode which effectively uses position again.

Well to be precise, the earlier change didn't really alter how we tracked anything, it just tightened up the declared types in the Java source (from `DiagnosticPosition` to `JCTree`) - the same actual objects are being used for tracking as before. So the code was "pretending" to use positions before, but it really was using trees.

> The fact that we have two ways of doing things worries me a bit -- if we want to keep using deferred lint handler in all places (both parser and after parsing) either it works on the lowest common denominator (the position), or we make some extra work in contexts where we only have a position but we need a tree.

I agree with the gist of what you're saying, i.e., Why have two ways of doing something?

So if I understand correctly, you and @lahodaj are suggesting the following:

1. Don't change how `DeferredLintHandler` tracks pending warnings, i.e., by their innermost containing `JCTree` declaration node.
1. Add another layer of "deferral", this one private to the parser, in which lexical warnings are deferred until their innermost `JCTree` declaration is created, at which time they can then be passed on to `DeferredLintHandler` like normal.

In my opinion Step 2 must be done in such a way that any future lexical warnings are able to "easily" use it, as we have several lexical lint warning feature requests queued up such as [JDK-8271171](https://bugs.openjdk.org/browse/JDK-8271171), [JDK-8210681](https://bugs.openjdk.org/browse/JDK-8210681), [JDK-6764629](https://bugs.openjdk.org/browse/JDK-6764629).

I will take a look at doing something like this in the parser. Basically, following Maurizio's steps.

We still have two ways of doing things but at least this way they'll be a little more compartmentalized... :)

Thanks for the comments.

mcimadamore commented:
> We still have two ways of doing things but at least this way they'll be a little more compartmentalized... :)

Well yes, but from the perspective of user model (and deferred lint handler) -- a lint warning would be _always_ be associated with some declaration. So it's up to the clients that have special requirements (such as parser and tokenizer) to make up for the difference -- rather than coming up with a more general way to map positions back to tree (which, if done, well at that point you could well be using positions everywhere).

Thanks for taking a look.

P.S.
I see a lot of changes related to positions -- @lahodaj @archiecobbs  do you thing that https://github.com/openjdk/jdk/pull/23669 would still be nice to have even if we don't need to map positions back to trees?

lahodaj commented:
I think I would prefer not to augment the trees with end positions until/unless we need them.

Regarding the new lint warnings - we probably should be careful with introducing new lints until we have some time to think about both conceptual and implementation directions for lints inside javac. (Although I admit I may have to introduce a new lint sometime soon.)

archiecobbs commented:
> rather than coming up with a more general way to map positions back to tree (which, if done, well at that point you could well be using positions everywhere)

Yes that was the original idea... i.e., eventually use positions everywhere, but "using positions everywhere" would require a larger refactoring of how warnings are handled that is a little too far in the future to assume right now.

> I think I would prefer not to augment the trees with end positions until/unless we need them.

Agreed - I'll plan to remove that part and close #23669 as part of the next update here.

> Regarding the new lint warnings - we probably should be careful with introducing new lints until we have some time to think about both conceptual and implementation directions for lints inside javac.

That's fine, right now I'm interested in simply making suppressible lexical warnings feasible. Actually adding (more of) them is a separate issue we can consider later.

vicente-romero-oracle commented:
<crazy-idea> would the design be simpler if we have a post-parser phase for this? </crazy-idea>

mcimadamore commented:
> `<crazy-idea>` would the design be simpler if we add a post-parser phase for this? `</crazy-idea>`

Not a crazy idea at all. Something similar was tried here:

https://github.com/openjdk/jdk/pull/23281#issuecomment-2640782274

An attempt was made to avoid deferral of lint warnings in general, and put all checks in the new pass we have. But it turned out to be more than we can chew -- at least in the sense that it's hard to get there in one single shot.

archiecobbs commented:
I've take a stab at the refactoring we discussed above. Now all of the work of mapping source code positions to `JCTree` declarations is handled entirely within the parser by a new class `LexicalLintManager`. Fire away :)

mcimadamore commented:
> I've take a stab at the refactoring we discussed above. Now all of the work of mapping source code positions to `JCTree` declarations is handled entirely within the parser by a new class `LexicalLintManager`. Fire away :)

Thanks for giving a try. I have an initial question about the code. It seems like it stores all the "reports" and all the "declarations" in flat lists. Then we flush the reports once per compilation unit.

This means we still have to do a lot of work to associate reports with declarations.

What I had in mind was to basically do the association when `endDecl` is called. E.g. once you see a field declaration, at that point you can defer all the lint warnings that have been created for that field. This probably needs some `startDecl` (maybe). I'm sure you considered this alternative -- I'm curious as to why it was rejected?

mcimadamore commented:
> > I've take a stab at the refactoring we discussed above. Now all of the work of mapping source code positions to `JCTree` declarations is handled entirely within the parser by a new class `LexicalLintManager`. Fire away :)
> 
> Thanks for giving a try. I have an initial question about the code. It seems like it stores all the "reports" and all the "declarations" in flat lists. Then we flush the reports once per compilation unit.
> 
> This means we still have to do a lot of work to associate reports with declarations.
> 
> What I had in mind was to basically do the association when `endDecl` is called. E.g. once you see a field declaration, at that point you can defer all the lint warnings that have been created for that field. This probably needs some `startDecl` (maybe). I'm sure you considered this alternative -- I'm curious as to why it was rejected?

Separately, I feel we're calling `endDecl` way more than what I had in mind. E.g. in my mental model the start/end of a "decl" is only relevant if it can have some suppression attached to it. I see the code calls out `endDecl` for things like binding variables in patterns and method parameters -- so I'm not sure I follow...

mcimadamore commented:
> E.g. in my mental model the start/end of a "decl" is only relevant if it can have some suppression attached to it.

What I mean is, one of these:
https://github.com/openjdk/jdk/blob/master/src/jdk.compiler/share/classes/com/sun/tools/javac/code/DeferredLintHandler.java#L119

mcimadamore commented:
> > E.g. in my mental model the start/end of a "decl" is only relevant if it can have some suppression attached to it.
> 
> What I mean is, one of these: https://github.com/openjdk/jdk/blob/master/src/jdk.compiler/share/classes/com/sun/tools/javac/code/DeferredLintHandler.java#L119

Ok, I think I get it -- e.g. stuff like deprecation warnings (since it's type-related) can be triggered by any variable declaration...
But... the parser doesn't know what types are, so IMHO, a lot of these declarations are not very interesting?

mcimadamore commented:
Some interesting (and extreme) cases:

```
class Test1 {
    @interface A {
        String value();
    }

    @SuppressWarnings("text-blocks")
    @A("""
                trailing space here:\u0020
                """)
    void m() { }
}
```

Or, even

```
class Test2 {
       @SuppressWarnings("""
                                           text-blocks\u0020
       """)
    void m() { }
}
```

I'm not saying we should handle them -- more pointing out that, perhaps, trying to come up with the "perfect"/"most general" scheme here might be very difficult (if at all possible).

mcimadamore commented:
> Some interesting (and extreme) cases:

The first actually works correctly with the changes in this PR (presumably because the start of the method decl includes the annotation). The second is just a buggy test I think -- if there's a space after the lint  category, I don't think javac can recognize it.

archiecobbs commented:
> Thanks for giving a try. I have an initial question about the code. It seems like it stores all the "reports" and all the "declarations" in flat lists. Then we flush the reports once per compilation unit.
> 
> This means we still have to do a lot of work to associate reports with declarations.

I originally was handling things "as they come", based on the assumption that reports always occur prior to the completion of their containing declaration, but that's not true (`dangling-doc-comment` violates it).

So I switched to storing everything then doing the match up at the end, but the flushing process still needs to be optimized to make it linear time.

> What I had in mind was to basically do the association when `endDecl` is called. E.g. once you see a field declaration, at that point you can defer all the lint warnings that have been created for that field. This probably needs some `startDecl` (maybe). I'm sure you considered this alternative -- I'm curious as to why it was rejected?

See previous comment - reports can come in "late", after their containing declaration. I did have a `startDecl` before but it was no longer needed once I added the logic to accommodate late reports.

> Separately, I feel we're calling endDecl way more than what I had in mind. E.g. in my mental model the start/end of a "decl" is only relevant if it can have some suppression attached to it. I see the code calls out endDecl for things like binding variables in patterns and method parameters -- so I'm not sure I follow...

First, we are only calling `endDecl` on declarations of: methods, packages, classes, variables, methods. I just added an assertion check similar to the one you linked to in 45c2aacff7a and it isn't firing so that's good.

As for _which_ variable declarations, we need to handle any that support `@SuppressWarnings` annotations, plain and simple - even if they can't have initializers. including variables in patterns and method parameters. The reason is because we may want to handle any future _lexical_ issues contained therein.

For example, after implementing [JDK-8271171](https://bugs.openjdk.org/browse/JDK-8271171), we would still want the following suppression to work:
```java
public record Example(int x) {

    public void method(Object f) {
        switch (f) {
        case Example(@SuppressWarnings("funky-unicode-escapes") int funkyVariable\u004Eame):
            break;
        default:
            break;
        }
    }
}
```

mcimadamore commented:
Thanks for the explanation. I think we'd probably would like to think some more on what is best to do here.

mcimadamore commented:
> I originally was handling things "as they come", based on the assumption that reports always occur prior to the completion of their containing declaration, but that's not true (`dangling-doc-comment` violates it).

Just one question: looking at your code, it seems that `saveDanglingDocComments` is always called _before_ `endDecl`. So I'm not sure how to read your comment. Perhaps you were hoping for all the warnings to be "contained" in the declaration body (and this would indeed be violated by comments, but also warnings on annotations).

mcimadamore commented:
> > I originally was handling things "as they come", based on the assumption that reports always occur prior to the completion of their containing declaration, but that's not true (`dangling-doc-comment` violates it).
> 
> Just one question: looking at your code, it seems that `saveDanglingDocComments` is always called _before_ `endDecl`. So I'm not sure how to read your comment. Perhaps you were hoping for all the warnings to be "contained" in the declaration body (and this would indeed be violated by comments, but also warnings on annotations).

Maybe you mean this:

```java
class Test {
   /* one */
   void test(/** two */ int i, /** three */ int y) { }
}
```

The two dangling warnings would be reported before the start of the method body -- but if we just flush things when we get to the closest `endDecl`, we'd report them as part of the method parameter declarations. Is that what you mean?

mcimadamore commented:
> The two dangling warnings would be reported before the start of the method body -- but if we just flush things when we get to the closest `endDecl`, we'd report them as part of the method parameter declarations. Is that what you mean?

Still, I'm confused because if I do this with the code in this PR:

```
class Test {
   /** one */
   void test(@SuppressWarnings("dangling-doc-comments") /** two */ int i, /** t>
}
```

One of the two warnings actually gets suppressed -- suggesting that the warning -- after all - is being attributed to the method parameter (which seems odd, given that method parameters cannot have javadoc comments).

archiecobbs commented:
> > I originally was handling things "as they come", based on the assumption that reports always occur prior to the completion of their containing declaration, but that's not true (`dangling-doc-comment` violates it).
> 
> Just one question: looking at your code, it seems that `saveDanglingDocComments` is always called _before_ `endDecl`. So I'm not sure how to read your comment.

Well I may just be confused. I saw that problem at the same time I was fighting with `VirtualParser`, so the "backwards" reports may have actually just been bogus duplicates.

In any case, it's a moot point now because the issue is no longer relevant in linear time refactoring (just pushed as 12d8c730c1f0051c52bd0d5362597ab0971867c7).

> One of the two warnings actually gets suppressed -- suggesting that the warning -- after all - is being attributed to the method parameter (which seems odd, given that method parameters cannot have javadoc comments).

I think the behavior of that example is correct. It doesn't matter if you put a Javadoc comment in front of something that doesn't support Javadoc comments - that just makes it definitely dangling, as opposed to maybe dangling!

The other thing playing into this example is that the lexical extent of variable declarations is defined to include any annotations on the declaration. Therefore the offending dangling Javadoc comment is contained by it, and therefore the suppression specified by the annotation applies. Therefore, the warning is suppressed.

archiecobbs commented:
Closing this PR. It's being made obsolete by #24584.
